File: python/ardu2igc/converter.py
"""Convert ArduPilot .bin logs to IGC format."""

from __future__ import annotations

import datetime
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

from .parser import ArdupilotLog, LogEntry
from .igc import IgcFix, IgcWriter, IgcHeader


@dataclass
class ConversionOptions:
    """Options for log conversion."""

    pilot_name: str = ""
    glider_type: str = ""
    glider_id: str = ""
    competition_id: str = ""
    competition_class: str = ""

    # Altitude source: "baro" uses BARO AltChr, "gps" uses GPS Alt, "both" uses baro for press, gps for gps
    altitude_source: str = "both"

    # Minimum GPS fix age to include (seconds). Filters out stale fixes.
    min_gps_age: float = 0.0

    # Maximum time gap between fixes (seconds). Inserts gaps in output.
    max_time_gap: float = 60.0

    # Fix accuracy override (meters). If 0, uses value from GPS message.
    fix_accuracy_override: int = 0

    # Use GPS2 message instead of GPS
    use_gps2: bool = False


def _gps_week_to_date(gps_week: int, gps_tow_ms: float) -> datetime.date:
    """Convert GPS week number and time-of-week to a UTC date."""
    # GPS epoch: January 6, 1980
    import datetime as _dt

    gps_epoch = _dt.date(1980, 1, 6)
    days = gps_week * 7 + int(gps_tow_ms / 86400000)
    return gps_epoch + _dt.timedelta(days=days)


def _time_ms_to_time(time_ms: float) -> datetime.time:
    """Convert milliseconds since midnight to a datetime.time."""
    total_secs = int(time_ms / 1000)
    hours = (total_secs // 3600) % 24
    minutes = (total_secs % 3600) // 60
    seconds = total_secs % 60
    return datetime.time(hours, minutes, seconds)


def _parse_gps_message(entry: LogEntry) -> Optional[IgcFix]:
    """Parse GPS or GPS2 message into an IgcFix."""
    fields = entry.fields

    # Latitude/longitude in degrees * 10^7
    lat_raw = fields.get("Lat")
    lon_raw = fields.get("Lng")
    if lat_raw is None or lon_raw is None:
        return None

    lat = float(lat_raw) / 1e7
    lon = float(lon_raw) / 1e7

    # Altitude
    alt = int(fields.get("Alt", 0))

    # Fix type: 0=no fix, 1=2D, 2=3D
    fix_type = fields.get("Status", fields.get("FixType", 0))
    valid = int(fix_type) >= 2

    # Time
    # GPS messages may have TimeMS (ms since midnight) or T (ms since week start)
    time_ms = fields.get("TimeMS") or fields.get("T")
    if time_ms is not None:
        ts = _time_ms_to_time(float(time_ms))
    else:
        ts = entry.timestamp_ms / 1000.0
        total_secs = int(ts) % 86400
        ts = datetime.time(
            (total_secs // 3600) % 24,
            (total_secs % 3600) // 60,
            total_secs % 60,
        )

    # Fix accuracy (HDop * ~5m or from message)
    hdop = fields.get("HDop", 1.0)
    if hdop is None or hdop <= 0:
        hdop = 1.0
    accuracy = int(float(hdop) * 5.0)

    return IgcFix(
        timestamp=ts,
        latitude=lat,
        longitude=lon,
        valid=valid,
        pressure_altitude=0,
        gps_altitude=alt,
        fix_accuracy=accuracy,
    )


def _parse_baro_message(entry: LogEntry) -> Optional[int]:
    """Parse BARO message, return pressure altitude in meters."""
    fields = entry.fields

    # BARO messages have Alt (barometric altitude)
    alt = fields.get("Alt")
    if alt is not None:
        return int(float(alt))

    # Some BARO variants use AltChr
    alt_chr = fields.get("AltChr")
    if alt_chr is not None:
        return int(float(alt_chr))

    return None


def _parse_gps_time(entry: LogEntry) -> Optional[tuple[int, float]]:
    """Extract GPS week and time-of-week from a GPS message."""
    fields = entry.fields
    week = fields.get("Week") or fields.get("GWk")
    time_ms = fields.get("TimeMS") or fields.get("T") or fields.get("GMS")

    if week and time_ms and int(week) > 0:
        return (int(week), float(time_ms))
    return None


def convert(
    input_path: str | Path,
    output_path: str | Path,
    options: Optional[ConversionOptions] = None,
) -> dict:
    """Convert an ArduPilot .bin log file to IGC format.

    Args:
        input_path: Path to the ArduPilot .bin log file.
        output_path: Path for the output .igc file.
        options: Conversion options.

    Returns:
        Dictionary with conversion statistics.
    """
    if options is None:
        options = ConversionOptions()

    log = ArdupilotLog(input_path)
    writer = IgcWriter()

    # Set header
    header_kwargs = {}
    if options.pilot_name:
        header_kwargs["pilot_name"] = options.pilot_name
    if options.glider_type:
        header_kwargs["glider_type"] = options.glider_type
    if options.glider_id:
        header_kwargs["glider_id"] = options.glider_id
    if options.competition_id:
        header_kwargs["competition_id"] = options.competition_id
    if options.competition_class:
        header_kwargs["competition_class"] = options.competition_class

    writer.set_header(**header_kwargs)

    # Try to extract flight date from GPS
    gps_time = log.get_first_gps_time()
    if gps_time:
        gps_week, _ = gps_time
        # Use today's date as fallback; actual date extraction needs GPS week calc
        pass

    # Collect barometric readings keyed approximately by time
    baro_data: dict[float, int] = {}
    gps_type = "GPS2" if options.use_gps2 else "GPS"

    # First pass: collect barometric data
    for entry in log.iter_messages("BARO", "BAR2"):
        ts = entry.timestamp_ms
        alt = _parse_baro_message(entry)
        if alt is not None:
            baro_data[ts] = alt

    # Second pass: process GPS fixes
    fix_count = 0
    skip_count = 0
    last_gps_time = None
    flight_date = None

    for entry in log.iter_messages(gps_type, "GPS"):
        gps_fix = _parse_gps_message(entry)
        if gps_fix is None:
            continue

        # Extract date from GPS week
        gps_info = _parse_gps_time(entry)
        if gps_info and flight_date is None:
            gps_week, gps_tow_ms = gps_info
            if gps_week > 0:
                flight_date = _gps_week_to_date(gps_week, gps_tow_ms)

        # Check time gap
        current_time = (
            gps_fix.timestamp.hour * 3600
            + gps_fix.timestamp.minute * 60
            + gps_fix.timestamp.second
        )
        if last_gps_time is not None:
            gap = current_time - last_gps_time
            if gap < 0:
                gap += 86400  # Handle midnight crossing
            if gap > options.max_time_gap:
                skip_count += 1
                last_gps_time = current_time
                continue

        # Find nearest barometric reading
        if options.altitude_source in ("baro", "both"):
            nearest_baro_ts = None
            min_diff = float("inf")
            for baro_ts in baro_data:
                diff = abs(baro_ts - entry.timestamp_ms)
                if diff < min_diff:
                    min_diff = diff
                    nearest_baro_ts = baro_ts

            if nearest_baro_ts is not None and min_diff < 5000:
                gps_fix.pressure_altitude = baro_data[nearest_baro_ts]
            else:
                # Use GPS altitude as fallback for pressure
                gps_fix.pressure_altitude = gps_fix.gps_altitude

        if options.altitude_source == "baro":
            gps_fix.gps_altitude = gps_fix.pressure_altitude
        elif options.altitude_source == "gps":
            gps_fix.pressure_altitude = gps_fix.gps_altitude

        # Apply fix accuracy override
        if options.fix_accuracy_override > 0:
            gps_fix.fix_accuracy = options.fix_accuracy_override

        writer.add_fix(gps_fix)
        fix_count += 1
        last_gps_time = current_time

    # Set flight date on header
    if flight_date:
        writer.set_header(date=flight_date)

    # Write output
    writer.write(output_path)

    return {
        "input": str(input_path),
        "output": str(output_path),
        "fixes_written": fix_count,
        "gaps_skipped": skip_count,
        "flight_date": str(flight_date) if flight_date else "unknown",
    }


def convert_info(input_path: str | Path) -> dict:
    """Get information about a .bin log file without converting.

    Returns:
        Dictionary with log file information.
    """
    log = ArdupilotLog(input_path)

    msg_counts = {}
    gps_info = None
    baro_count = 0

    for entry in log.iter_messages():
        mtype = entry.msg_type
        msg_counts[mtype] = msg_counts.get(mtype, 0) + 1

        if mtype in ("GPS", "GPS2") and gps_info is None:
            gps_info = _parse_gps_time(entry)

    for entry in log.iter_messages("BARO", "BAR2"):
        baro_count += 1

    return {
        "file": str(input_path),
        "message_types": len(msg_counts),
        "total_messages": sum(msg_counts.values()),
        "message_counts": dict(sorted(msg_counts.items())),
        "has_gps": "GPS" in msg_counts or "GPS2" in msg_counts,
        "has_baro": baro_count > 0,
        "baro_readings": baro_count,
        "gps_info": gps_info,
    }

File: csharp/Ardu2IGC/Converter.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Ardu2IGC;

/// <summary>
/// Options for converting ArduPilot .bin logs to IGC format.
/// </summary>
public sealed class ConversionOptions
{
    public string PilotName { get; set; } = "";
    public string GliderType { get; set; } = "";
    public string GliderId { get; set; } = "";
    public string CompetitionId { get; set; } = "";
    public string CompetitionClass { get; set; } = "";

    /// <summary>Altitude source: "baro", "gps", or "both".</summary>
    public string AltitudeSource { get; set; } = "both";

    /// <summary>Maximum time gap between fixes in seconds before inserting a break.</summary>
    public double MaxTimeGap { get; set; } = 60.0;

    /// <summary>Override fix accuracy in meters. 0 = auto from GPS message.</summary>
    public int FixAccuracyOverride { get; set; } = 0;

    /// <summary>Use GPS2 message instead of GPS.</summary>
    public bool UseGps2 { get; set; } = false;
}

/// <summary>
/// Result of a conversion operation.
/// </summary>
public sealed class ConversionResult
{
    public string Input { get; init; } = "";
    public string Output { get; init; } = "";
    public int FixesWritten { get; init; }
    public int GapsSkipped { get; init; }
    public string FlightDate { get; init; } = "unknown";
}

/// <summary>
/// Converts ArduPilot .bin log files to IGC format.
/// </summary>
public static class ArdupilotConverter
{
    private static readonly DateTime GpsEpoch = new(1980, 1, 6, 0, 0, 0, DateTimeKind.Utc);

    /// <summary>
    /// Convert an ArduPilot .bin log file to IGC format.
    /// </summary>
    public static ConversionResult Convert(
        string inputPath,
        string outputPath,
        ConversionOptions? options = null)
    {
        options ??= new ConversionOptions();

        var log = new ArdupilotLog(inputPath);
        var writer = new IgcWriter();

        // Set header
        if (!string.IsNullOrEmpty(options.PilotName))
            writer.Header.PilotName = options.PilotName;
        if (!string.IsNullOrEmpty(options.GliderType))
            writer.Header.GliderType = options.GliderType;
        if (!string.IsNullOrEmpty(options.GliderId))
            writer.Header.GliderId = options.GliderId;
        if (!string.IsNullOrEmpty(options.CompetitionId))
            writer.Header.CompetitionId = options.CompetitionId;
        if (!string.IsNullOrEmpty(options.CompetitionClass))
            writer.Header.CompetitionClass = options.CompetitionClass;

        // Collect barometric data
        var baroData = new SortedDictionary<double, int>();
        foreach (var entry in log.IterateMessages("BARO", "BAR2"))
        {
            if (entry.Fields.TryGetValue("Alt", out var altVal))
            {
                baroData[entry.TimestampMs] = System.Convert.ToInt32(altVal);
            }
            else if (entry.Fields.TryGetValue("AltChr", out var altChrVal))
            {
                baroData[entry.TimestampMs] = System.Convert.ToInt32(altChrVal);
            }
        }

        // Process GPS fixes
        string gpsType = options.UseGps2 ? "GPS2" : "GPS";
        int fixCount = 0;
        int skipCount = 0;
        double? lastGpsTime = null;
        DateTime? flightDate = null;

        foreach (var entry in log.IterateMessages(gpsType, "GPS"))
        {
            var gpsFix = ParseGpsMessage(entry);
            if (gpsFix is null) continue;

            // Extract flight date from GPS week
            if (flightDate is null)
            {
                var gpsTime = ExtractGpsTime(entry);
                if (gpsTime.HasValue)
                {
                    flightDate = GpsEpoch.AddDays(gpsTime.Value.Week * 7)
                        .AddMilliseconds(gpsTime.Value.TimeMs);
                }
            }

            // Check time gap
            double currentTime = gpsFix.Timestamp.Hour * 3600
                + gpsFix.Timestamp.Minute * 60
                + gpsFix.Timestamp.Second;

            if (lastGpsTime.HasValue)
            {
                double gap = currentTime - lastGpsTime.Value;
                if (gap < 0) gap += 86400; // Handle midnight crossing
                if (gap > options.MaxTimeGap)
                {
                    skipCount++;
                    lastGpsTime = currentTime;
                    continue;
                }
            }

            // Find nearest barometric reading
            int pressAlt = gpsFix.GpsAltitude;
            if (options.AltitudeSource is "baro" or "both")
            {
                double minDiff = double.MaxValue;
                int nearestBaro = 0;
                foreach (var kvp in baroData)
                {
                    double diff = Math.Abs(kvp.Key - entry.TimestampMs);
                    if (diff < minDiff)
                    {
                        minDiff = diff;
                        nearestBaro = kvp.Value;
                    }
                }
                if (minDiff < 5000)
                    pressAlt = nearestBaro;
            }

            int gpsAlt = options.AltitudeSource == "baro"
                ? pressAlt
                : gpsFix.GpsAltitude;

            if (options.AltitudeSource == "gps")
                pressAlt = gpsAlt;

            int fxa = options.FixAccuracyOverride > 0
                ? options.FixAccuracyOverride
                : gpsFix.FixAccuracy;

            writer.AddFix(new IgcFix
            {
                Timestamp = gpsFix.Timestamp,
                Latitude = gpsFix.Latitude,
                Longitude = gpsFix.Longitude,
                Valid = gpsFix.Valid,
                PressureAltitude = pressAlt,
                GpsAltitude = gpsAlt,
                FixAccuracy = fxa,
            });

            fixCount++;
            lastGpsTime = currentTime;
        }

        // Set flight date on header
        if (flightDate.HasValue)
            writer.Header.Date = flightDate.Value.Date;

        // Write output
        writer.Write(outputPath);

        return new ConversionResult
        {
            Input = inputPath,
            Output = outputPath,
            FixesWritten = fixCount,
            GapsSkipped = skipCount,
            FlightDate = flightDate?.ToString("yyyy-MM-dd") ?? "unknown",
        };
    }

    /// <summary>
    /// Get information about a .bin log file without converting.
    /// </summary>
    public static Dictionary<string, object> GetInfo(string inputPath)
    {
        var log = new ArdupilotLog(inputPath);
        var msgCounts = new Dictionary<string, int>();
        int baroCount = 0;
        int? gpsWeek = null;

        foreach (var entry in log.IterateMessages())
        {
            if (!msgCounts.ContainsKey(entry.MsgType))
                msgCounts[entry.MsgType] = 0;
            msgCounts[entry.MsgType]++;

            if (entry.MsgType is "GPS" or "GPS2" && gpsWeek is null)
            {
                if (entry.Fields.TryGetValue("Week", out var w) && System.Convert.ToInt32(w) > 0)
                    gpsWeek = System.Convert.ToInt32(w);
                else if (entry.Fields.TryGetValue("GWk", out var gwk) && System.Convert.ToInt32(gwk) > 0)
                    gpsWeek = System.Convert.ToInt32(gwk);
            }
        }

        foreach (var entry in log.IterateMessages("BARO", "BAR2"))
            baroCount++;

        int totalMsgs = 0;
        foreach (var c in msgCounts.Values) totalMsgs += c;

        return new Dictionary<string, object>
        {
            ["file"] = inputPath,
            ["messageTypes"] = msgCounts.Count,
            ["totalMessages"] = totalMsgs,
            ["hasGps"] = msgCounts.ContainsKey("GPS") || msgCounts.ContainsKey("GPS2"),
            ["hasBaro"] = baroCount > 0,
            ["baroReadings"] = baroCount,
            ["gpsWeek"] = gpsWeek ?? 0,
        };
    }

    private static IgcFix? ParseGpsMessage(LogEntry entry)
    {
        if (!entry.Fields.TryGetValue("Lat", out var latRaw) ||
            !entry.Fields.TryGetValue("Lng", out var lonRaw))
            return null;

        double lat = System.Convert.ToDouble(latRaw) / 1e7;
        double lon = System.Convert.ToDouble(lonRaw) / 1e7;
        int alt = entry.Fields.TryGetValue("Alt", out var altVal)
            ? System.Convert.ToInt32(altVal) : 0;

        int fixType = entry.Fields.TryGetValue("Status", out var status)
            ? System.Convert.ToInt32(status)
            : entry.Fields.TryGetValue("FixType", out var fixType2)
                ? System.Convert.ToInt32(fixType2) : 0;
        bool valid = fixType >= 2;

        // Time
        DateTime timestamp;
        if (entry.Fields.TryGetValue("TimeMS", out var timeMs))
        {
            double ms = System.Convert.ToDouble(timeMs);
            int totalSecs = (int)(ms / 1000) % 86400;
            timestamp = new DateTime(2000, 1, 1,
                totalSecs / 3600, (totalSecs % 3600) / 60, totalSecs % 60,
                DateTimeKind.Utc);
        }
        else
        {
            int totalSecs = (int)(entry.TimestampMs / 1000) % 86400;
            timestamp = new DateTime(2000, 1, 1,
                totalSecs / 3600, (totalSecs % 3600) / 60, totalSecs % 60,
                DateTimeKind.Utc);
        }

        // Fix accuracy from HDop
        int accuracy = 50;
        if (entry.Fields.TryGetValue("HDop", out var hdopVal))
        {
            double hdop = System.Convert.ToDouble(hdopVal);
            if (hdop > 0)
                accuracy = (int)(hdop * 5.0);
        }

        return new IgcFix
        {
            Timestamp = timestamp,
            Latitude = lat,
            Longitude = lon,
            Valid = valid,
            GpsAltitude = alt,
            FixAccuracy = accuracy,
        };
    }

    private static (int Week, double TimeMs)? ExtractGpsTime(LogEntry entry)
    {
        int week = 0;
        double timeMs = 0;

        if (entry.Fields.TryGetValue("Week", out var w))
            week = System.Convert.ToInt32(w);
        else if (entry.Fields.TryGetValue("GWk", out var gwk))
            week = System.Convert.ToInt32(gwk);

        if (entry.Fields.TryGetValue("TimeMS", out var tms))
            timeMs = System.Convert.ToDouble(tms);
        else if (entry.Fields.TryGetValue("T", out var t))
            timeMs = System.Convert.ToDouble(t);
        else if (entry.Fields.TryGetValue("GMS", out var gms))
            timeMs = System.Convert.ToDouble(gms);

        if (week > 0 && timeMs > 0)
            return (week, timeMs);
        return null;
    }
}
